File: cpp/03_two_sum_III.cpp
//Question:
//Design and implement a TwoSum class. It should support the following
//operations: add and find.
//add(input) – Add the number input to an internal data structure.
//find(value) – Find if there exists any pair of numbers which sum is
//equal to the value.
//
//For example,
//add(1); add(3); add(5); find(4) - true; find(7) - false
#include <iostream>
#include <vector>
using namespace std;

class TwoSum {
private:
    vector<int> data;
public:
    void add(int number) {
        data.insert(number);
    }
    bool find(int value) {
        unordered_map<int, int> hash;
        for (int i = 0; i < data.size(); i++) {
            int numberToFind = target - data[i];
            //if numberToFind is found in map, return them
            if (hash.find(numberToFind) != hash.end()) {
                return true;
            }
            //number was not found. Put it in the map.
            hash[data[i]] = i;
        }
        return false;
    }
};

File: python/03_two_sum_III.py
"""
Question:
Design and implement a TwoSum class. It should support the following
operations: add and find.
add(input) – Add the number input to an internal data structure.
find(value) – Find if there exists any pair of numbers which sum is
equal to the value.

For example,
add(1); add(3); add(5); find(4) - true; find(7) - false
"""


class TwoSum:
    def __init__(self):
        self.structure = []

    def add(self, input):
        self.structure.append(input)

    def find(self, value):
        return self.twoSum(self.structure, value)

    @staticmethod
    def twoSum(nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hash_map = {}
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in hash_map:
                return True  # [hash_map[complement], i]
            else:
                hash_map[nums[i]] = i
        return False
